### PR TITLE
Improve `parse_expr` and use in `process_dimension_pragmas`

### PR DESCRIPTION
**Pull request by MichaelSt98**
Improve `parse_expr` and use in `process_dimension_pragmas`.

* add correct parsing of `:` to `RangeIndex((None, None))`
* support `DerivedType` s

### CODE REVIEW

**Comment by MichaelSt98** [comment]
Some improvements to the expression parser included in this PR:

Executing:

```py
from loki import parse_expr

class Foo:
    val3 = 1
    arr = [[1, 2], [3, 4]]
    def __init__(self, _val1, _val2):
        self.val1 = _val1
        self.val2 = _val2
    def some_func(self, a, b):
        return a + b
    @staticmethod
    def static_func(a):
        return 2*a

context = {'foo': Foo(2, 3)}
test_str = 'foo%val1 + foo%val2 + foo%val3'
parsed = parse_expr(f'{test_str}')
print(f"parsing: {test_str} with context: '{context}'\n  results in '{parsed}'")
parsed = parse_expr(f'{test_str}', evaluate=True, context=context)
print(f"parsing and evaluating: {test_str} with context: '{context}'\n  results in '{parsed}'")
test_str = 'foo%val1 + foo%some_func(1, 2) + foo%static_func_2(3)'
parsed = parse_expr(f'{test_str}', evaluate=True, context=context)
print(f"parsing and evaluating: {test_str} with context: '{context}'\n  results in '{parsed}'")
test_str = 'foo%val1 + foo%some_func(1, 2) + foo%static_func(3) + foo%arr(1, 2)'
parsed = parse_expr(f'{test_str}', evaluate=True, context=context, strict=True)
print(f"parsing and evaluating: {test_str} with context: '{context}'\n  results in '{parsed}'")
test_str = 'foo%val1 + foo%some_func(1, b=2) + foo%static_func(a=3) + foo%arr(1, 2)'
parsed = parse_expr(f'{test_str}', evaluate=True, context=context, strict=True)
print(f"parsing and evaluating: {test_str} with context: '{context}'\n  results in '{parsed}'")
```

gives

```
parsing: foo%val1 + foo%val2 + foo%val3 with context: '{'foo': <__main__.Foo object at 0x1494ff19c3d0>}'
  results in 'foo%val1 + foo%val2 + foo%val3'
parsing and evaluating: foo%val1 + foo%val2 + foo%val3 with context: '{'foo': <__main__.Foo object at 0x1494ff19c3d0>}'
  results in '6'
parsing and evaluating: foo%val1 + foo%some_func(1, 2) + foo%static_func_2(3) with context: '{'foo': <__main__.Foo object at 0x1494ff19c3d0>}'
  results in '5 + foo%static_func_2(3)'
parsing and evaluating: foo%val1 + foo%some_func(1, 2) + foo%static_func(3) + foo%arr(1, 2) with context: '{'foo': <__main__.Foo object at 0x1494ff19c3d0>}'
  results in '13'
parsing and evaluating: foo%val1 + foo%some_func(1, b=2) + foo%static_func(a=3) + foo%arr(1, 2) with context: '{'foo': <__main__.Foo object at 0x1494ff19c3d0>}'
  results in '13'
```

**Review by reuterbal — CHANGES_REQUESTED** [review]
Many thanks, the improvements to `parse_expr` and how it can deal with derived types is absolutely fantastic! I'm really impressed by the degree of functionality you have created in a short amount of time.

Just out of curiosity but not necessarily something we need: would the evaluation also work for nested derived types? Say something like

```python
class Bar:
    baz = 4

class Foo:
    bar = Bar()

parse_expr('1 + foo%bar%baz', context={'foo': Foo()}, evaluate=True)
```

I am, however, not convinced by the integration in `process_dimension_pragmas`. The workaround with the expanded regex pattern breaks down as soon as there's more than one set of nested parentheses, and it's fundamentally a problem that cannot be solved by regex matching.

Therefore, I would propose to re-implement the `get_pragma_parameters` utility to provide this functionality. But we can also do this as a separate PR and leave out the support for parentheses in this PR.

File: loki/expression/tests/test_expression.py [hunk 1]
```diff
@@ -2064,6 +2101,11 @@ def test_expression_parser_evaluate(case):
     parsed = parse_expr(convert_to_case(f'{test_str}', mode=case), evaluate=True, context=context)
     assert parsed == 1
 
+    context = {'arr': [[1, 2], [3, 4]]}
+    test_str = '1 + arr(1, 2)'
+    parsed = parse_expr(convert_to_case(f'{test_str}', mode=case), evaluate=True, context=context)
+    assert parsed == 3
+
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply loki/expression/tests/test_expression.py:2109]: As discussed offline ...

File: loki/ir/pragma_utils.py [hunk 2]
```diff
@@ -88,13 +97,22 @@ def get_pragma_parameters(pragma, starts_with=None, only_loki_pragmas=True):
...
             content = content[len(starts_with):]
-        for match in re.finditer(_get_pragma_parameters_re, content):
+        for match in re.finditer(pattern, content):
             parameters[match.group('command')].append(match.group('arg'))
     parameters = {k: v if len(v) > 1 else v[0] for k, v in parameters.items()}
     return parameters
 
 
-def process_dimension_pragmas(ir):
+_get_pragma_dim_parameter_re = re.compile(r'(?P<command>[\w-]+)\s*(?:\((?P<arg>.*)\))?')
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply loki/ir/pragma_utils.py]: I agree. I did what you suggested based on the link you provided.

**Comment by MichaelSt98** [comment]
Regarding your nested derived types question:
It was not possible, but I've added a commit to allow nested derived types. 

With this, executing:

```py
from loki import parse_expr

class Bar:
    baz = 4

class Foo:
    bar = Bar()

ir = parse_expr('1 + foo%bar%baz', context={'foo': Foo()}, evaluate=True)
print(ir)
```

yields `5`.

There are also more sophisticated tests for nested derived types in `expression/tests/test_expression.py`

File: loki/expression/parser.py [hunk 3]
```diff
@@ -201,8 +241,46 @@ def map_call(self, expr):
...
+            if isinstance(expr.name, pmbl.CallWithKwargs):
+                name = expr.name.function.name
+                args = [self.rec(par) for par in expr.name.parameters]
+                kwargs = {
+                    k: self.rec(v)
+                    for k, v in expr.name.kw_parameters.items()}
+                kwargs = CaseInsensitiveDict(kwargs)
+                return self.case_insensitive_getattr(self.rec(expr.aggregate), name)(*args, **kwargs)
+
+        except Exception as e:
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply loki/expression/parser.py:285]: Hmm not sure how to narrow this down ... I want to catch any and all exceptions, possibly but not necessarily thrown by pymbolic (as some parent methods are called) ... Isn't `except Exception` exactly what to do in this case?

[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/expression/parser.py:285]: Well, the purpose should be to gracefully handle only situations that you consider "safe" in this particular context - so, in this context likely situations like incomplete evaluation due to undefined variables/functions? Can these not be attributed to more specific exception classes. Unexpected exceptions should fall through verbatim and be dealt with at an outer level in my opinion.

[the remark this answers is not in the document]
Comment (MichaelSt98) [reply loki/expression/parser.py:285]: I see your point. However, I (re)throw the exception if `self.strict = True` and the whole logic is to `try` what I want you to do e.g., evaluate the expression and if you fail for whatever reason just return the `ir`/`expression` (unless I want to have the evaluated expression and if this is not possible, throw the `exception` and tell me why this is not possible)?!

File: loki/expression/parser.py [hunk 4]
```diff
@@ -255,30 +255,39 @@ def map_call_with_kwargs(self, expr):
...
+                    (k, self.rec(v))
+                    for k, v in current_expr.name.kw_parameters.items()
+                )
+                return obj(*args, **kwargs)
         except Exception as e:
             if self.strict:
                 raise e
             return expr
+        if self.strict:
+            raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/expression/parser.py:290]: Sorry, you're right of course. I missed the fact that we won't reach this point if it's one of the known "lookup" nodes. Thanks for clarifying!

**Review by reuterbal — APPROVED** [review]
Excellent, many thanks!

**Pull request merged**